packages/grafana-ui/src/components/Pagination/Pagination.tsx

```diff
@@ -16,21 +19,57 @@ export const Pagination: React.FC<Props> = ({ currentPage, numberOfPages, onNavi
   const styles = getStyles();
   const pages = [...new Array(numberOfPages).keys()];
 
+  const condensePages = numberOfPages >= PAGE_LENGTH_TO_CONDENSE;
+  const getListItem = (page: number, variant: 'primary' | 'secondary') => (
+    <li key={page} className={styles.item}>
+      <Button size="sm" variant={variant} onClick={() => onNavigate(page)}>
+        {page}
+      </Button>
+    </li>
+  );
+
   return (
     <div className={styles.container}>
       <ol>
-        {pages.map(pageIndex => {
+        <li className={styles.item}>
+          <Button
+            size="sm"
+            variant="secondary"
+            onClick={() => onNavigate(currentPage - 1)}
+            disabled={currentPage === 1}
+          >
+            <Icon name="angle-left" />
+          </Button>
+        </li>
+        {pages.reduce<JSX.Element[]>((pagesToRender, pageIndex) => {
           const page = pageIndex + 1;
           const variant: ButtonVariant = page === currentPage ? 'primary' : 'secondary';
 
-          return (
-            <li key={page} className={styles.item}>
-              <Button size="sm" variant={variant} onClick={() => onNavigate(page)}>
-                {page}
-              </Button>
-            </li>
-          );
-        })}
+          if (condensePages) {
+            if (page === 1 || page === numberOfPages || (page >= currentPage - 2 && page <= currentPage + 2)) {
+              pagesToRender.push(getListItem(page, variant));
+            } else if (page === currentPage - 3 || page === currentPage + 3) {
+              pagesToRender.push(
+                <li key={page} className={styles.item}>
+                  <Icon className={styles.ellipsis} name="ellipsis-v" />
+                </li>
+              );
+            }
```

Comment:
I think this logic is causing this fairly weird behavior of the navigation arrows changing it's position when navigating, as in the gif below:

![Uploading Kapture 2021-01-13 at 15.00.53.gif…]()

Reply:
Regarding this comment - maybe @jessover9000 or @hemdrup could chime in and provide some more feedback from the UX POV?

Reply:
Yes, the jumping buttons are strange. If we want as little as possible jumping when paging through, we need to make sure we always have the same count of elements, so if we have one … separator we need one outer button + 7 buttons on the other side,  or for two … separators two outer buttons and 5 middle buttons. If we think that's a bit too many buttons or we need to make it work for narrow panel widths, we can also just use 3 middle buttons instead of 5, but it should be an odd number so that the active page is centered.
Concretely for this example that would mean:

- Beginning of pagination: Buttons from 1–7 + … + last number
- Somewhere in the middle: 1 + … + 2 previous buttons + 1x current highlighted number + 2 next buttons + … + last number
- End of pagination: 1 + … + 6 buttons prior to last number + last number

Hope it's clear what I mean 🤪

Reply:
This is proving to be a challenging logic problem. I've drawn out a few scenarios here that I'll try to implement

This is using 5 middle buttons, with a total length of elements including navigation buttons at 11

<  1  2  3  ***4***  5  6  7 ...  10  >
<  1 ...  4  5  ***6***  7  8  9  10  >
<  1 ...  8  9  ***10***  11  12 ...  15  >

If it were to be 3 middle buttons it would look something like this

<  1  2  3  ***4***  5  ...  10  >
<  1 ...  5  ***6***  7  ...  10  >
<  1 ...  8  ***9*** 10 >